### PR TITLE
Input Validation for Create Report Definition

### PR DESCRIPTION
**Pull request by davidcui-amzn**
*Issue #, if available:*
#90 
*Description of changes:*
Added input validation on required fields from the client side on `Create report definition`
* Added requirements for `Report name` abiding by the allowed characters listed in the field 
* Added requirements for the interval to populated with a number when `Recurring By interval` is selected
* Added cron validation for `Cron-based` schedule definitions
* Added requirements that the list of email recipients cannot be empty when delivering reports by email

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: kibana-reports/public/components/report_definitions/create/create_report_definition.tsx [hunk 1]
```diff
@@ -150,6 +171,50 @@ export function CreateReport(props) {
...
+      if (
+        !isValidCron(metadata.trigger.trigger_params.schedule.cron.expression)
+      ) {
+        setShowCronError(true);
+        error = true;
+      }
+    }
+    // if email delivery and no recipients are listed
+    if (metadata.delivery.delivery_type === 'Channel') {
+      if (metadata.delivery.delivery_params.recipients.length === 0) {
```
Comment (zhongnansu) [inline kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:210]: can you add some basic check for email string?, e.g. it has to be in the format of `xxx@xxx`

File: kibana-reports/public/components/report_definitions/create/create_report_definition.tsx [hunk 2]
```diff
@@ -164,33 +229,43 @@ export function CreateReport(props) {
...
-        console.log('error in creating report definition: ' + error + message);
-        handleErrorToast();
-      });
+    let error = false;
+    await definitionInputValidation(metadata, error).then((response) => {
+      error = response;
+    });
+    if (error) {
+      handleErrorToast();
+      setPreErrorData(metadata);
```
Comment (zhongnansu) [inline kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:238]: Why do we need this?

**Review by zhongnansu — APPROVED** [review]
Please address the comments

File: kibana-reports/public/components/report_definitions/create/create_report_definition.tsx [hunk 3]
```diff
@@ -150,6 +171,50 @@ export function CreateReport(props) {
...
+      ) {
+        setShowCronError(true);
+        error = true;
+      }
+    }
+    // if email delivery and no recipients are listed
+    if (metadata.delivery.delivery_type === 'Channel') {
+      if (metadata.delivery.delivery_params.recipients.length === 0) {
+        setShowEmailRecipientsError(true);
+        error = true;
```
[the remark this answers is not in the document]
Comment (davidcui-amzn) [reply kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:212]: Sure

File: kibana-reports/public/components/report_definitions/create/create_report_definition.tsx [hunk 4]
```diff
@@ -150,6 +171,50 @@ export function CreateReport(props) {
...
+      }
+    }
+
+    // if cron based and cron input is invalid
+    if (
+      metadata.trigger.trigger_type === 'Schedule' &&
+      metadata.trigger.trigger_params.schedule_type === 'Cron based'
+    ) {
+      if (
+        !isValidCron(metadata.trigger.trigger_params.schedule.cron.expression)
```
[the remark this answers is not in the document]
Comment (davidcui-amzn) [reply kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:202]: Yeah it is- I tested it with `30 11 * * *` which is accepted- the library validates against Unix cron.

File: kibana-reports/public/components/report_definitions/create/create_report_definition.tsx (hunk 2 above)
Comment (davidcui-amzn) [reply kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:238]: If the user submits an invalid report definition, then the values will all reset in the client-side due to how `createReportDefinitionRequest` is declared and initialized. This saves the state of the data before the submit, and assigns it to `createReportDefinitionRequest` so the values persist

Comment (zhongnansu) [reply kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:249]: So does that mean the input validation only happens when user submit the report, not after he fills in each field?

Comment (davidcui-amzn) [reply kibana-reports/public/components/report_definitions/create/create_report_definition.tsx:249]: Correct- the `EuiFormRow` has an `isInvalid` flag but it's only set `onSubmit` right now

**Review by abbashus — APPROVED** [review]

**Review by abbashus — APPROVED** [review]

**Pull request merged**